### PR TITLE
[SQL][SPARK-17490] Optimize SerializeFromObject() for a primitive array

### PR DESCRIPTION
**Pull request by kiszk**
## What changes were proposed in this pull request?

Waiting for merging #13680

This PR optimizes `SerializeFromObject()` for an primitive array. This is derived from #13758 to address one of problems by using a simple way in #13758.

The current implementation always generates `GenericArrayData` from `SerializeFromObject()` for any type of an array in a logical plan. This involves a boxing at a constructor of `GenericArrayData` when `SerializedFromObject()` has an primitive array.

This PR enables to generate `UnsafeArrayData` from `SerializeFromObject()` for a primitive array. It can avoid boxing to create an instance of `ArrayData` in the generated code by Catalyst.

This PR also generate `UnsafeArrayData` in a case for `RowEncoder.serializeFor` or `CatalystTypeConverters.createToCatalystConverter`.

Performance improvement of `SerializeFromObject()` is up to 2.0x

```
OpenJDK 64-Bit Server VM 1.8.0_91-b14 on Linux 4.4.11-200.fc22.x86_64
Intel Xeon E3-12xx v2 (Ivy Bridge)

Without this PR
Write an array in Dataset:               Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
Int                                            556 /  608         15.1          66.3       1.0X
Double                                        1668 / 1746          5.0         198.8       0.3X

with this PR
Write an array in Dataset:               Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
Int                                            352 /  401         23.8          42.0       1.0X
Double                                         821 /  885         10.2          97.9       0.4X
```

Here is an example program that will happen in mllib as described in [SPARK-16070](https://issues.apache.org/jira/browse/SPARK-16070).

```
sparkContext.parallelize(Seq(Array(1, 2)), 1).toDS.map(e => e).show
```

Generated code before applying this PR

``` java
/* 039 */   protected void processNext() throws java.io.IOException {
/* 040 */     while (inputadapter_input.hasNext()) {
/* 041 */       InternalRow inputadapter_row = (InternalRow) inputadapter_input.next();
/* 042 */       int[] inputadapter_value = (int[])inputadapter_row.get(0, null);
/* 043 */
/* 044 */       Object mapelements_obj = ((Expression) references[0]).eval(null);
/* 045 */       scala.Function1 mapelements_value1 = (scala.Function1) mapelements_obj;
/* 046 */
/* 047 */       boolean mapelements_isNull = false || false;
/* 048 */       int[] mapelements_value = null;
/* 049 */       if (!mapelements_isNull) {
/* 050 */         Object mapelements_funcResult = null;
/* 051 */         mapelements_funcResult = mapelements_value1.apply(inputadapter_value);
/* 052 */         if (mapelements_funcResult == null) {
/* 053 */           mapelements_isNull = true;
/* 054 */         } else {
/* 055 */           mapelements_value = (int[]) mapelements_funcResult;
/* 056 */         }
/* 057 */
/* 058 */       }
/* 059 */       mapelements_isNull = mapelements_value == null;
/* 060 */
/* 061 */       serializefromobject_argIsNulls[0] = mapelements_isNull;
/* 062 */       serializefromobject_argValue = mapelements_value;
/* 063 */
/* 064 */       boolean serializefromobject_isNull = false;
/* 065 */       for (int idx = 0; idx < 1; idx++) {
/* 066 */         if (serializefromobject_argIsNulls[idx]) { serializefromobject_isNull = true; break; }
/* 067 */       }
/* 068 */
/* 069 */       final ArrayData serializefromobject_value = serializefromobject_isNull ? null : new org.apache.spark.sql.catalyst.util.GenericArrayData(serializefromobject_argValue);
/* 070 */       serializefromobject_holder.reset();
/* 071 */
/* 072 */       serializefromobject_rowWriter.zeroOutNullBytes();
/* 073 */
/* 074 */       if (serializefromobject_isNull) {
/* 075 */         serializefromobject_rowWriter.setNullAt(0);
/* 076 */       } else {
/* 077 */         // Remember the current cursor so that we can calculate how many bytes are
/* 078 */         // written later.
/* 079 */         final int serializefromobject_tmpCursor = serializefromobject_holder.cursor;
/* 080 */
/* 081 */         if (serializefromobject_value instanceof UnsafeArrayData) {
/* 082 */           final int serializefromobject_sizeInBytes = ((UnsafeArrayData) serializefromobject_value).getSizeInBytes();
/* 083 */           // grow the global buffer before writing data.
/* 084 */           serializefromobject_holder.grow(serializefromobject_sizeInBytes);
/* 085 */           ((UnsafeArrayData) serializefromobject_value).writeToMemory(serializefromobject_holder.buffer, serializefromobject_holder.cursor);
/* 086 */           serializefromobject_holder.cursor += serializefromobject_sizeInBytes;
/* 087 */
/* 088 */         } else {
/* 089 */           final int serializefromobject_numElements = serializefromobject_value.numElements();
/* 090 */           serializefromobject_arrayWriter.initialize(serializefromobject_holder, serializefromobject_numElements, 4);
/* 091 */
/* 092 */           for (int serializefromobject_index = 0; serializefromobject_index < serializefromobject_numElements; serializefromobject_index++) {
/* 093 */             if (serializefromobject_value.isNullAt(serializefromobject_index)) {
/* 094 */               serializefromobject_arrayWriter.setNullInt(serializefromobject_index);
/* 095 */             } else {
/* 096 */               final int serializefromobject_element = serializefromobject_value.getInt(serializefromobject_index);
/* 097 */               serializefromobject_arrayWriter.write(serializefromobject_index, serializefromobject_element);
/* 098 */             }
/* 099 */           }
/* 100 */         }
/* 101 */
/* 102 */         serializefromobject_rowWriter.setOffsetAndSize(0, serializefromobject_tmpCursor, serializefromobject_holder.cursor - serializefromobject_tmpCursor);
/* 103 */       }
/* 104 */       serializefromobject_result.setTotalSize(serializefromobject_holder.totalSize());
/* 105 */       append(serializefromobject_result);
/* 106 */       if (shouldStop()) return;
/* 107 */     }
/* 108 */   }
/* 109 */ }
```

Generated code after applying this PR

``` java
/* 035 */   protected void processNext() throws java.io.IOException {
/* 036 */     while (inputadapter_input.hasNext()) {
/* 037 */       InternalRow inputadapter_row = (InternalRow) inputadapter_input.next();
/* 038 */       int[] inputadapter_value = (int[])inputadapter_row.get(0, null);
/* 039 */
/* 040 */       Object mapelements_obj = ((Expression) references[0]).eval(null);
/* 041 */       scala.Function1 mapelements_value1 = (scala.Function1) mapelements_obj;
/* 042 */
/* 043 */       boolean mapelements_isNull = false || false;
/* 044 */       int[] mapelements_value = null;
/* 045 */       if (!mapelements_isNull) {
/* 046 */         Object mapelements_funcResult = null;
/* 047 */         mapelements_funcResult = mapelements_value1.apply(inputadapter_value);
/* 048 */         if (mapelements_funcResult == null) {
/* 049 */           mapelements_isNull = true;
/* 050 */         } else {
/* 051 */           mapelements_value = (int[]) mapelements_funcResult;
/* 052 */         }
/* 053 */
/* 054 */       }
/* 055 */       mapelements_isNull = mapelements_value == null;
/* 056 */
/* 057 */       boolean serializefromobject_isNull = mapelements_isNull;
/* 058 */       final ArrayData serializefromobject_value = serializefromobject_isNull ? null : org.apache.spark.sql.catalyst.expressions.UnsafeArrayData.fromPrimitiveArray(mapelements_value);
/* 059 */       serializefromobject_isNull = serializefromobject_value == null;
/* 060 */       serializefromobject_holder.reset();
/* 061 */
/* 062 */       serializefromobject_rowWriter.zeroOutNullBytes();
/* 063 */
/* 064 */       if (serializefromobject_isNull) {
/* 065 */         serializefromobject_rowWriter.setNullAt(0);
/* 066 */       } else {
/* 067 */         // Remember the current cursor so that we can calculate how many bytes are
/* 068 */         // written later.
/* 069 */         final int serializefromobject_tmpCursor = serializefromobject_holder.cursor;
/* 070 */
/* 071 */         if (serializefromobject_value instanceof UnsafeArrayData) {
/* 072 */           final int serializefromobject_sizeInBytes = ((UnsafeArrayData) serializefromobject_value).getSizeInBytes();
/* 073 */           // grow the global buffer before writing data.
/* 074 */           serializefromobject_holder.grow(serializefromobject_sizeInBytes);
/* 075 */           ((UnsafeArrayData) serializefromobject_value).writeToMemory(serializefromobject_holder.buffer, serializefromobject_holder.cursor);
/* 076 */           serializefromobject_holder.cursor += serializefromobject_sizeInBytes;
/* 077 */
/* 078 */         } else {
/* 079 */           final int serializefromobject_numElements = serializefromobject_value.numElements();
/* 080 */           serializefromobject_arrayWriter.initialize(serializefromobject_holder, serializefromobject_numElements, 4);
/* 081 */
/* 082 */           for (int serializefromobject_index = 0; serializefromobject_index < serializefromobject_numElements; serializefromobject_index++) {
/* 083 */             if (serializefromobject_value.isNullAt(serializefromobject_index)) {
/* 084 */               serializefromobject_arrayWriter.setNullInt(serializefromobject_index);
/* 085 */             } else {
/* 086 */               final int serializefromobject_element = serializefromobject_value.getInt(serializefromobject_index);
/* 087 */               serializefromobject_arrayWriter.write(serializefromobject_index, serializefromobject_element);
/* 088 */             }
/* 089 */           }
/* 090 */         }
/* 091 */
/* 092 */         serializefromobject_rowWriter.setOffsetAndSize(0, serializefromobject_tmpCursor, serializefromobject_holder.cursor - serializefromobject_tmpCursor);
/* 093 */       }
/* 094 */       serializefromobject_result.setTotalSize(serializefromobject_holder.totalSize());
/* 095 */       append(serializefromobject_result);
/* 096 */       if (shouldStop()) return;
/* 097 */     }
/* 098 */   }
/* 099 */ }
```
## How was this patch tested?

Added a test in `DatasetSuite`, `RowEncoderSuite`, and `CatalystTypeConvertersSuite`

### CODE REVIEW

**Comment by hvanhovell** [comment]
@kiszk this seems nice, is this ready for review?

**Comment by hvanhovell** [comment]
retest this please

**Comment by kiszk** [comment]
Thanks, I should update one file and update it later today in Japan.

PR #13758 can also solve this issue without allocating UnsafeArrayData. I think that PR #13758 is a generic solution and has small amount of changes. Which PR is preferable, #15044 or #13758?

**Comment by kiszk** [comment]
@hvanhovell this PR is ready for review. Which PR is preferable, #15044 or #13758?

**Comment by hvanhovell** [comment]
@kiszk this PR is. It is much more concise!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/PrimitiveArrayBenchmark.scala [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+/**
+ * Benchmark [[PrimitiveArray]] for DataFrame and Dataset program using primitive array
+ * To run this:
+ *  1. replace ignore(...) with test(...)
+ *  2. build/sbt "sql/test-only *benchmark.PrimitiveArrayDataBenchmark"
+ *
+ * Benchmarks in this file are skipped in normal builds.
+ */
+class PrimitiveArrayBenchmark extends BenchmarkBase {
```
Comment (hvanhovell) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/PrimitiveArrayBenchmark.scala]: Could you post some of the results?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala [hunk 2]
```diff
@@ -441,6 +441,28 @@ object ScalaReflection extends ScalaReflection {
           val newPath = s"""- array element class: "$clsName"""" +: walkedTypePath
           MapObjects(serializerFor(_, elementType, newPath), input, dt)
 
+         case dt @ (BooleanType | ByteType | ShortType | IntegerType | LongType |
```
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala]: I am afraid that I cannot understand your question correctly.
This case covers only cases that we are interested in (e.g. generate `UnsafeArrayData`). Other cases are covered by [here](https://github.com/apache/spark/pull/15044/files/df300c534670e65cb44ae570859711382f85043b#diff-89643554d9757dd3e91abff1cc6096c7R466).

Is this an answer to you?

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala]: I mean that there a few more places in which we can apply this:

- https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala#L125-L128
- https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala#L161-L172

Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala]: Thank you for your clarification. Let me check them.

**Comment by kiszk** [comment]
@hvanhovell, after my investigations, I have added code to generate `UnsafeArrayData' at two code paths. Could you please review this again?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystTypeConvertersSuite.scala [hunk 3]
```diff
@@ -61,4 +63,39 @@ class CatalystTypeConvertersSuite extends SparkFunSuite {
   test("option handling in createToCatalystConverter") {
     assert(CatalystTypeConverters.createToCatalystConverter(IntegerType)(Some(123)) === 123)
   }
+
+  test("primitive array handing") {
```
Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystTypeConvertersSuite.scala]: Nit: handing => handling

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystTypeConvertersSuite.scala [hunk 4]
```diff
@@ -61,4 +63,39 @@ class CatalystTypeConvertersSuite extends SparkFunSuite {
...
+    val doubleArray = Array(1.1, 111.1, 11111.1)
+    val doubleUnsafeArray = UnsafeArrayData.fromPrimitiveArray(doubleArray)
+    val doubleArrayType = ArrayType(DoubleType, false)
+    assert(CatalystTypeConverters.createToScalaConverter(doubleArrayType)(doubleUnsafeArray)
+      === doubleArray)
+    assert(CatalystTypeConverters.createToCatalystConverter(doubleArrayType)(doubleArray)
+      == doubleUnsafeArray)
+  }
+
+  test("An array with null handing") {
```
Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystTypeConvertersSuite.scala]: Nit: handing => handling

**Comment by kiszk** [comment]
Jenkins, retest this please

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystTypeConvertersSuite.scala (hunk 4 above)
Comment (kiszk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystTypeConvertersSuite.scala]: Good catch. Addressed both of them.

**Comment by kiszk** [comment]
@hvanhovell could you please review this again?

**Comment by kiszk** [comment]
Jenkins, retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala [hunk 5]
```diff
@@ -119,18 +119,36 @@ object RowEncoder {
...
-      case _ => MapObjects(
-        element => serializerFor(ValidateExternalType(element, et), et),
-        inputObject,
-        ObjectType(classOf[Object]))
-    }
+    case t @ ArrayType(et, cn) =>
+      val cls = inputObject.dataType.asInstanceOf[ObjectType].cls
+      et match {
+/*
+        case BooleanType | ByteType | ShortType | IntegerType | LongType | FloatType | DoubleType
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala]: What is going on here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala [hunk 6]
```diff
@@ -193,7 +211,8 @@ object RowEncoder {
     // as java.lang.Object.
     case _: DecimalType => ObjectType(classOf[java.lang.Object])
     // In order to support both Array and Seq in external row, we make this as java.lang.Object.
-    case _: ArrayType => ObjectType(classOf[java.lang.Object])
+    case a @ ArrayType(et, cn) =>
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala]: Why this change?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 7]
```diff
@@ -993,7 +993,7 @@ case class ValidateExternalType(child: Expression, expected: DataType)
       case _: DecimalType =>
         Seq(classOf[java.math.BigDecimal], classOf[scala.math.BigDecimal], classOf[Decimal])
           .map(cls => s"$obj instanceof ${cls.getName}").mkString(" || ")
-      case _: ArrayType =>
+      case a @ ArrayType(et, cn) =>
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: Same here: why this change?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala [hunk 8]
```diff
@@ -158,6 +158,20 @@ object CatalystTypeConverters {
...
+        case a: Array[Short] =>
+          UnsafeArrayData.fromPrimitiveArray(a)
+        case a: Array[Int] =>
+          UnsafeArrayData.fromPrimitiveArray(a)
+        case a: Array[Long] =>
+          UnsafeArrayData.fromPrimitiveArray(a)
+        case a: Array[Float] =>
+          UnsafeArrayData.fromPrimitiveArray(a)
+        case a: Array[Double] =>
+          UnsafeArrayData.fromPrimitiveArray(a)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala]: do we need to do this? The `CatalystTypeConverter` will be removed and replaced by encoder eventually. Does your benchmark cover this branch?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala [hunk 9]
```diff
@@ -441,6 +441,28 @@ object ScalaReflection extends ScalaReflection {
...
+                    FloatType | DoubleType) =>
+          val cls = input.dataType.asInstanceOf[ObjectType].cls
+          if (cls.isAssignableFrom(classOf[Array[Boolean]]) ||
+              cls.isAssignableFrom(classOf[Array[Byte]]) ||
+              cls.isAssignableFrom(classOf[Array[Short]]) ||
+              cls.isAssignableFrom(classOf[Array[Int]]) ||
+              cls.isAssignableFrom(classOf[Array[Long]]) ||
+              cls.isAssignableFrom(classOf[Array[Float]]) ||
+              cls.isAssignableFrom(classOf[Array[Double]])) {
+            StaticInvoke(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala]: we can simplify it 
```
if (cls.isArray && cls.getComponentType.isPrimitive) ...
```
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala [hunk 10]
```diff
@@ -119,18 +119,36 @@ object RowEncoder {
         "fromString",
         inputObject :: Nil)
 
-    case t @ ArrayType(et, _) => et match {
-      case BooleanType | ByteType | ShortType | IntegerType | LongType | FloatType | DoubleType =>
-        // TODO: validate input type for primitive array.
-        NewInstance(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala]: can we do the same thing here? i.e. special handling primitive array. I know we don't have the class information here, bu we can do it in the runtime:
```
object ArrayData {
  def toArrayData(input: Any): ArrayData = input match {
    case a: Array[Boolean] => UnsafeArrayData.fromPrimitive(a)
    ...
    case other => new GenericArrayData(other)
  }
}
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala (hunk 8 above)
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala]: My benchmark does not cover this branch.
[This test](https://github.com/kiszk/spark/blob/b5473e3b6675c6bbfdd3fe261699f855faf39c94/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystTypeConvertersSuite.scala#L72-L73) causes an failure without this branch. Should we drop this test and this branch?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala (hunk 10 above)
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala]: Good idea. It works well.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala (hunk 5 above)
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala]: Sorry, my mistake

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala (hunk 6 above)
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala]: I reverted it, but it was not the same as the previous one. thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala (hunk 7 above)
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala (hunk 9 above)
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala]: thanks, good catch.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala (hunk 8 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala]: yea let's remove them, `CatalystTypeConverter` will not be in critical path and we don't need to optimize it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala [hunk 11]
```diff
@@ -119,18 +119,20 @@ object RowEncoder {
...
-        inputObject,
-        ObjectType(classOf[Object]))
-    }
+    case t @ ArrayType(et, cn) =>
+      val cls = inputObject.dataType.asInstanceOf[ObjectType].cls
+      et match {
+        case BooleanType | ByteType | ShortType | IntegerType | LongType | FloatType | DoubleType =>
+          StaticInvoke(
+            classOf[ArrayData],
+            ObjectType(classOf[ArrayData]),
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala]: why not just use the given ArrayType `t` here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala [hunk 12]
```diff
@@ -119,18 +119,20 @@ object RowEncoder {
...
-          classOf[GenericArrayData],
-          inputObject :: Nil,
-          dataType = t)
-      case _ => MapObjects(
-        element => serializerFor(ValidateExternalType(element, et), et),
-        inputObject,
-        ObjectType(classOf[Object]))
-    }
+    case t @ ArrayType(et, cn) =>
+      val cls = inputObject.dataType.asInstanceOf[ObjectType].cls
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala]: where do we use the `cls`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/RowEncoderSuite.scala [hunk 13]
```diff
@@ -191,6 +191,32 @@ class RowEncoderSuite extends SparkFunSuite {
     assert(encoder.serializer.head.nullable == false)
   }
 
+  test("RowEncoder should support a primitive array") {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/RowEncoderSuite.scala]: nit: `RowEncoder should support primitive arrays`

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 14]
```diff
@@ -987,6 +987,24 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
       checkAnswer(agg, ds.groupBy('id % 2).agg(count('id)))
     }
   }
+
+  test("array") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: can you think of a better test name which can describe what we are testing here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala (hunk 11 above)
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala]: yes, done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala (hunk 12 above)
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala]: oh, removed

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala (hunk 14 above)
Comment (kiszk) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: good catch, gave a better name

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/RowEncoderSuite.scala (hunk 13 above)
Comment (kiszk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/RowEncoderSuite.scala]: done

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/PrimitiveArrayBenchmark.scala [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+
+  def writeDatasetArray(iters: Int): Unit = {
+    import sparkSession.implicits._
+
+    val count = 1024 * 1024 * 2
+
+    val sc = sparkSession.sparkContext
+    val primitiveIntArray = Array.fill[Int](count)(65535)
+    val dsInt = sc.parallelize(Seq(primitiveIntArray), 1).toDS
+    dsInt.count
```
Comment (kiszk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/PrimitiveArrayBenchmark.scala]: Want to force to build dataset

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/PrimitiveArrayBenchmark.scala [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+    val intArray = { i: Int =>
+      var n = 0
+      while (n < iters) {
+        dsInt.map(e => e).collect
```
Comment (kiszk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/PrimitiveArrayBenchmark.scala]: I see. done

**Comment by cloud-fan** [comment]
LGTM, can you also address https://github.com/apache/spark/pull/15044#discussion_r86665473?

**Comment by kiszk** [comment]
Sorry for missing to commit [this](https://github.com/kiszk/spark/blob/4c679b59b34ea01474baa5d4ac4ce89133a1720d/sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala#L991).

**Comment by hvanhovell** [comment]
Merging to master/2.1. Thanks for the hard work!
